src/identifier.rs

```diff
@@ -1,14 +1,42 @@
 //! Define the type of an identifier.
+use once_cell::sync::Lazy;
 use serde::{Deserialize, Serialize};
-use std::{fmt, hash::Hash};
+use std::fmt;
+use std::hash::Hash;
 
 use crate::position::TermPos;
 
-#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+static INTERNER: Lazy<interner::Interner> = Lazy::new(interner::Interner::new);
+
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
 #[serde(into = "String", from = "String")]
 pub struct Ident {
-    pub label: String,
-    pub pos: TermPos,
+    label: interner::Symbol,
+    pos: TermPos,
+    generated: bool,
+}
+
+impl Ident {
+    pub fn new_with_pos(label: impl AsRef<str>, pos: TermPos) -> Self {
+        let generated = label.as_ref().starts_with(GEN_PREFIX);
```

Comment:
I wonder if we couldn't do something a bit cleaner here, by avoiding allocating a string at all for generated symbols, by:

- having a method to print a symbol as a number (`Symbol::to_string()`)
- move the method to create a fresh identifiers to `Ident`, which would set `generated` to `true` in this case
- if `label()` is called, just return `format!("{}{}", GEN_PREFIX, symbol`.

I can see that the last one may be problematic because `label` return a `&str`. But at least we can move the code to create a fresh generated identifier into `Ident` and handle `generated` in a less fragile way (rather than relying on the presence of some character).

Reply:
Not sure this is relative to this PR, but it could be a nice improvement, definitely :)